src/activity_queue.rs

```diff
@@ -220,27 +248,413 @@ pub(crate) fn generate_request_headers(inbox_url: &Url) -> HeaderMap {
+        worker_count: usize,
+        retry_count: usize,
+        timeout: Duration,
+        backoff: usize, // This should be 60 seconds by default or 1 second in tests
+    ) -> Self {
+        let stats: Arc<Stats> = Default::default();
+
+        // This task clears the dead/completed stats every hour
+        let hour_stats = stats.clone();
+        tokio::spawn(async move {
+            let duration = Duration::from_secs(3600);
+            loop {
+                tokio::time::sleep(duration).await;
+                hour_stats.completed_last_hour.store(0, Ordering::Relaxed);
+                hour_stats.dead_last_hour.store(0, Ordering::Relaxed);
+            }
+        });
+
+        let (retry_sender, mut retry_receiver) = unbounded_channel();
+        let retry_stats = stats.clone();
+        let retry_client = client.clone();
+
+        // The "fast path" retry
+        // The backoff should be < 5 mins for this to work otherwise signatures may expire
+        // This strategy is the one that is used with the *same* signature
+        let strategy = RetryStrategy {
+            backoff,
+            retries: 1,
+            offset: 0,
+            initial_sleep: 0,
+        };
+
+        // The "retry path" strategy
+        // After the fast path fails, a task will sleep up to backoff ^ 2 and then retry again
+        let retry_strategy = RetryStrategy {
+            backoff,
+            retries: 3,
+            offset: 2,
+            initial_sleep: backoff.pow(2), // wait 60 mins before even trying
+        };
+
+        let retry_sender_task = tokio::spawn(async move {
+            let semaphore = Arc::new(Semaphore::new(retry_count));
+            let mut join_set = JoinSet::new();
+
+            while let Some(message) = retry_receiver.recv().await {
+                let permit = semaphore
+                    .clone()
+                    .acquire_owned()
+                    .await
+                    .expect("should never be closed");
+
+                join_set.spawn(retry_worker(
+                    retry_client.clone(),
+                    timeout,
+                    message,
+                    permit,
+                    retry_stats.clone(),
+                    retry_strategy,
+                ));
+            }
+
+            while !join_set.is_empty() {
+                join_set.join_next().await;
+            }
+        });
+
+        let (sender, mut receiver) = unbounded_channel();
+
+        let sender_stats = stats.clone();
+
+        let sender_task = tokio::spawn(async move {
+            let semaphore = Arc::new(Semaphore::new(worker_count));
+            let mut join_set = JoinSet::new();
+
+            while let Some(message) = receiver.recv().await {
+                let permit = semaphore
+                    .clone()
+                    .acquire_owned()
+                    .await
+                    .expect("should never be closed");
+
+                join_set.spawn(worker(
```

Comment:
for the semaphore to work correctly, I think you have to move the permit into the inner spawn call so it gets released only after the worker has finished. (`join_set.spawn(async move { let _permit = permit; worker().await; })`).

But also, I don't think you need the semaphore at all and instead should rather do what I did in my patch: Wait for tasks to finish if there's more than the limit within the inner loop (before accepting the next one):

```rust
                join_set.spawn(...);
                while join_set.len() > worker_count {
                    // prevent there being more than worker_count running tasks and remove finished ones from join_set.
                    join_set.join_next().await;
                }
```

This should fulfill the same purpose as the semaphore but also prevent join_set from just keep accumulating tasks (even those that are already finished). Ideally it would just automatically remove the finished ones but I didn't see any option for that in joinset and I don't know if there's a better way.

Reply:
This has been resolved! Now using the `join_set` to apply backpressure rather than semaphores